### DISCUSSION TITLE
[Help]: Feedback and advice on a carousel use case with animations

### DISCUSSION THREAD

**Discussion by stepanjakl** [davidjerleke/embla-carousel#334] [Help]
Hello good people,

Firstly, thank you @davidjerleke for a great library! I am absolutely in love with it.

This issues is not necessarily a bug report or feature request - more like a feedback for both sides.

I need to use Embla in a specific use case or format. I believe that I have been pretty successful with it so far.

Let me show and tell you, what it is about.

<img width="1015" alt="Screenshot 2022-05-25 at 12 36 09" src="https://user-images.githubusercontent.com/47005900/170253254-451417d8-f9b2-48ec-a2a4-5f161b4d4d26.png">

<img width="1069" alt="Screenshot 2022-05-25 at 12 34 42" src="https://user-images.githubusercontent.com/47005900/170253270-ef41e96b-e47d-46b2-9cd8-306c916f4887.png">

In an essence, I need slides that come outside the container of the carousel to stack under each other - on both sides. As the slides move "off the limits" they animate `translate3d` and `scale` transforms as well as an `opacity` property.

Nothing difficult, you would think, but still a day of work went into to get it just right.

Now, the Embla library is nicely written without a doubt. However, figuring out its "inner workings" is not a simple task. Specifically when it comes to the `internalEngine()` function and general structure of the API (as there is no real documentation). I struggled most with getting the right snap values in place i.e. `scrollSnapList()` function and "tying" with the `scrollProgress()`. My understanding is that these two functions are the core calculations for getting the slide positioning and subsequently their transitions/animations right and smooth.

I was wondering if you could take a look at the code and provide a feedback on the code and overall implementation. I am curious to see if there is a more cleaner or effective way of getting to the above use case of Embla carousel. 

Here is the code: https://codesandbox.io/s/h7bkx1?file=/src/js/index.js

A few notes:
- ~~The script is unfortunately not dynamic when slide width changes i.e. `.embla__slide` CSS class and `slidesPerView` variable (only works with 2 slides in the "middle" at the moment)~~ **UPDATE - fixed now**
- Setting `containScroll` to `trimSnaps` breaks it
- I am quite unsure if `oneSnap` variable to calculate "distance" between two slides is correctly implemented, maybe there is a better approach?
- Is there an engine helper for calculating when the slide is off limits (i.e. `leftOffset`/`rightOffset` vars in my case)?
- It isn't working correctly when the `loop` option is enabled in the current configuration. Moreover, I noticed when a container has `grid` display property instead of `flex`, the loop functionality is totally broken.
- Any other suggestions and/or concerns would be greatly welcome.

I would be very happy for you to use it as one of the examples for the official documentation once the code is a bit polished.

If possible, I would like to request a documentation for the `internalEngine()` function :)

**Comment by davidjerleke** [comment]
Hi @stepanjakl,

Thank you for sharing this. It looks like you're almost there 🙂. As much as I would love to help everyone out with their feature requests I can't. My spare time is very limited and I would say that your request can be categorized as extending the carousel beyond its core functionality. Unfortunately, this category doesn't have the highest priority right now. 

At the time of writing I'm working on the stable v7 release and that is pretty much eating up the little spare time I have. And if you add all that needs to be done with new examples, updating the documentation and so on you can image that it's quite a lot to handle for me.

> If possible, I would like to request a documentation for the internalEngine() function :)

Actually, not adding `internalEngine()` to the documentation is intentional. As you mention:

> Now, the Embla library is nicely written without a doubt. However, figuring out its "inner workings" is not a simple task.

It basically exposes the inner functionality of Embla and all its utilities which makes the library very very extensible. And very useful for plugin authors. But in order to use it properly one has to understand its "inner workings". Documenting this method would lead to users creating tons of bugs that aren't actual bugs but rather stuff they break because they simply don't understand how it works. I simply don't have time and can't handle that flood of "false" bugs. In contrast, the "public" API is designed to not break as easily.

> I would be very happy for you to use it as one of the examples for the official documentation once the code is a bit polished.

I would happily add it to the docs as soon as it's finished 👍. I can assist with skimming the code and look for improvements once it's done.

> - Is there an engine helper for calculating when the slide is off limits (i.e. leftOffset/rightOffset vars in my case)?

The first method that comes to mind is `slidesInView()` and `slidesNotInView()` but these might not work as expected when altering slide positions with transforms.

> - It isn't working correctly when the loop option is enabled in the current configuration. Moreover, I noticed when a container has grid display property instead of flex, the loop functionality is totally broken.

I'd say make sure that your CSS with the grid setup is correct. 99% of the time there's something wrong with the CSS setup. This is because Embla makes minimal DOM changes in order to create a carousel. It reads slide positions and acts on that so if your slides aren't aligned along the scroll axis within the container it won't work properly.
However, there are bugs that are solved in the new [release candidate](https://github.com/davidjerleke/embla-carousel/releases) v7.0.0-rc01 that might be related to this problem:

- #285
- #317

I hope this helps.

Cheers,
David

**Reply by stepanjakl** [reply]
Hi @davidjerleke,

Thank you for such a great and detailed answer!

I agree with exposing the `internalEngine()` function, it makes sense that it could create more problems then solve. I absolutely understand that you're stretched for time. Time is a scarce commodity these days haha. Embla is a great tool, better than other carousels out there (imho). Its simplicity and size makes it a real challenger. My only wish is to expose more functions to allow making custom animations easier - such as the one we're discussing.

Anyway, I decided to discard the original proposed solution in the Sandbox attached. I was getting lost in the code and there were many little issues that I kept patching up. I think the main issue stemmed from working with float snap values (`scrollSnapValues()`)  and translating/interpolating them to pixel values. It just became difficult to do the math.

Instead, I opted for working with DOMRect object.

```
const animateSlides = () => {
      const animateSlides = () => {
    this.embla.slideNodes().forEach((slide, index) => {
      let overflowSide = null
      let distanceFromLeftSide = -(sliderViewport.getBoundingClientRect().left - slide.getBoundingClientRect().left) let distanceFromRightSide = sliderViewport.getBoundingClientRect().right - slide.getBoundingClientRect().right
      let slideWidth = slide.getBoundingClientRect().width
      if (distanceFromLeftSide < 0) {
        overflowSide = 'left'
        slide.firstElementChild.style.transformOrigin = 'center'
        slide.firstElementChild.style.transform = `translate3d(${everpolate.linear(-distanceFromLeftSide, [ 0, slideWidth * 2 ], [ 0, slideWidth * (1 + 1 / 2) ])}px, 0px, 0px) scale(${everpolate.linear(-distanceFromLeftSide, [ 0, slideWidth * 2 ], [ 1, (3 / 4) ])})`
        slide.firstElementChild.style.opacity = everpolate.linear(-distanceFromLeftSide, [0, slideWidth * 2], [1, 0]) slide.style.zIndex = 0
      }
      else if (distanceFromRightSide < 0) {
        overflowSide = 'right'
        .
        .
      }
      else {
        slide.firstElementChild.style.transformOrigin = null slide.firstElementChild.style.transform = `translate3d(0px, 0px, 0px) scale(1)`
        slide.firstElementChild.style.opacity = 1 slide.style.zIndex = 1
      }
    })
```

This way I got a better control over the input values and resulting animations.

However, I kept having an issue with "jiggering" slides, the animation was just not smooth enough. And indeed there was a discrepancy between the Embla values and actual DOMRect values during scroll - likely to do with how Embla handles dragging. In any case, I am very close to completion (with a more legible code) as I discovered the `internalEngine().target.get()` which gives me the right value. From my understanding, it is used for the final translate3d transform on the container value. It gives off the final value, but not the intermediate values on pointerUp. Is there any way I can retrieve the animated translate3d value? If not, I'll just regex it from the actual DOM I think 😄 

Excerpt from the latest solution:

```
let slideWidth = slide.getBoundingClientRect().width

let distanceFromLeftSide = engine.target.get() + (index * slideWidth)
let distanceFromRightSide = engine.target.get() + sliderViewport.getBoundingClientRect().width - ((index + 1) * slideWidth)

let distanceFromLeftSideProgress = everpolate.linear(-distanceFromLeftSide, [ 0, slideWidth ], [ 0, 1 ])[0]
slide.firstElementChild.style.transform = `translate3d(${everpolate.linear(distanceFromLeftSideProgress, [ 0, 1 ], [ 0, 100 ])[0]}%, 0px, 0px)`
```

What do you think?

(It would be really nice if there were more position values exposed for each slide e.g. distance from snap points and distance from being out of view - in pixels.)

**Reply by davidjerleke** [reply]
Hello again @stepanjakl,

### 🚨 **Note!**
Before continuing with what you want to achieve I highly recommend you to work with the latest release candidate for v7. The current candidate is [v7.0.0-rc04](https://github.com/davidjerleke/embla-carousel/releases/tag/v7.0.0-rc04). The main reason for this is because Embla v7 is using `px` when transforming its container instead of `%`. This will make your life easier when working with the `embla.internalEngine()` method because you don't need to convert values from `px` to `%` and vice versa. Every value exposed there is in pixels.

---

With that out of the way, the `embla.internalEngine()` method exposes the following: 

<img width="506" alt="internalEngine" src="https://user-images.githubusercontent.com/11529148/177986986-430ed609-7a85-4c6f-94f1-08c4973f9943.png">

Now I can't explain what all of these methods/properties do but I'll try to pick some of them that might be relevant for your use case:

```ts
const engine = embla.internalEngine()
```

Get the current/intermediate transform value in `px`:

```ts
engine.location.get()
```

Get the target transform value in `px`, which is the final transform value where the carousel is headed:
```ts
engine.target.get()
```

A list of all scroll snap points where the carousel will snap into position in `px`. Note that if your content is `ltr` (left to right which is the default [direction](https://www.embla-carousel.com/api/options/#direction) value), most of these will be negative because scrolling forward means moving the container to the left:

```ts
engine.scrollSnaps
```

Scroll limit which is the start bound and end bound of the carousel. For a `ltr` and a horizontal carousel, which is the most common setup, the left side is the start bound and the right side is the end bound. These might not be relevant for a carousel with loop enabled:

```ts
engine.limit.max // start bound
engine.limit.min // end bound
engine.limit.length // the scrollable length of the carousel from max --> min
```

SlideRects - The DOM rect state of the slides stored when the carousel was initialized. These values won't update if you were to transform your slides. They will only update if you hard reset the carousel, which happens automatically when the window is resized and Embla internally calls `embla.reInit()`, or when you manually run `embla.reInit()`. You can use these to calculate where the slides are because you have access to the slide rects and the container rect plus the container transform value with `engine.location.get()`. And if you transform your slides you can add that offset to each slide when calculating stuff:

```ts
engine.slideRects
```

Now exposing more than `embla.internalEngine()` already does might not make sense. This is because it will increase bundle size and might lead to Embla exposing more stuff than it actually needs to fulfill its purpose. I hope that makes sense.

Best,
David

**Reply by stepanjakl** [reply]
Thank you David, that's super helpful. I'll come back with the final solution once it's done :)

**Reply by stepanjakl** [reply]
Here it is: https://codesandbox.io/s/embla-carousel-custom-animation-vanilla-2ergmw

I am pretty happy with it now. Only notion I have is regarding the layout, I just couldn't make it work with `grid`. It doesn't seem to take into consideration the `grid-column-gap` property and overflows the container as a consequence.

Otherwise, what do you think of the code?

**Reply by davidjerleke** [reply]
Hi @stepanjakl,

That was fast! I like the folding effect so we should definitely add this to the docs. I’ll check the code out and get back to you when possible. A lot has to be prepared for the stable v7 release so please be patient.

Cheers,
David
